test/integration/Utils.kt

```diff
@@ -45,152 +46,200 @@ import java.nio.file.Path
+    const val SCHEMA_FILE_NAME = "schema_string.tql"
+
+    const val FAIL_CONNECT_TYPEDB = "Failed to connect to TypeDB."
+    const val FAIL_CREATE_DATABASE = "Failed to create the database."
+    const val FAIL_DATA_WRITE = "Failed to write the data."
+    const val FAIL_SCHEMA_WRITE = "Failed to write the schema."
+
+    fun runComposeRule(compose: ComposeContentTestRule, rule: suspend ComposeContentTestRule.() -> Unit) {
+        runBlocking { compose.rule() }
+    }
 
+    fun studioTest(compose: ComposeContentTestRule, funcBody: suspend () -> Unit) {
+        runComposeRule(compose) {
+            setContent {
+                Studio.MainWindowContent(WindowContext.Test(1000, 1000, 0, 0))
+            }
+            funcBody()
+        }
+}
 
-val SAVE_ICON_STRING = Icon.Code.FLOPPY_DISK.unicode
-val PLUS_ICON_STRING = Icon.Code.PLUS.unicode
-val PLAY_ICON_STRING = Icon.Code.PLAY.unicode
-val CHECK_ICON_STRING = Icon.Code.CHECK.unicode
-val ROLLBACK_ICON_STRING = Icon.Code.ROTATE_LEFT.unicode
-val CHEVRON_UP_ICON_STRING = Icon.Code.CHEVRON_UP.unicode
-val DOUBLE_CHEVRON_DOWN_ICON_STRING = Icon.Code.CHEVRONS_DOWN.unicode
-val DOUBLE_CHEVRON_UP_ICON_STRING = Icon.Code.CHEVRONS_UP.unicode
+    fun studioTestWithRunner(compose: ComposeContentTestRule, funcBody: suspend (String) -> Unit) {
+        val typeDB = TypeDBCoreRunner()
+        typeDB.start()
+        val address = typeDB.address()
+        runComposeRule(compose) {
+            setContent {
+                Studio.MainWindowContent(WindowContext.Test(1000, 1000, 0, 0))
+            }
+            funcBody(address)
+        }
+        typeDB.stop()
+    }
 
-val SAMPLE_DATA_PATH = File("test/data/sample_file_structure").absolutePath
-val TQL_DATA_PATH = File("test/data").absolutePath
+    fun cloneAndOpenProject(composeRule: ComposeContentTestRule, source: String, destination: String): Path {
+        val absolute = File(File(destination).absolutePath)
 
-const val QUERY_FILE_NAME = "query_string.tql"
-const val DATA_FILE_NAME = "data_string.tql"
-const val SCHEMA_FILE_NAME = "schema_string.tql"
+        absolute.deleteRecursively()
+        File(source).copyRecursively(overwrite = true, target = absolute)
 
-const val DB_ADDRESS = "localhost:1729"
+        StudioState.project.tryOpenProject(absolute.toPath())
 
-fun runComposeRule(compose: ComposeContentTestRule, rule: suspend ComposeContentTestRule.() -> Unit) {
-    runBlocking { compose.rule() }
-}
+        composeRule.waitForIdle()
+        return absolute.toPath()
+    }
 
-fun studioTest(compose: ComposeContentTestRule, funcBody: suspend () -> Unit) {
-    runComposeRule(compose) {
-        setContent {
-            Studio.MainWindowContent(WindowContext.Test(1000, 1000, 0, 0))
-        }
-        funcBody()
+    /// Wait `timeMillis` milliseconds, then wait for all recompositions to finish.
+    suspend fun waitAndRecompose(composeRule: ComposeContentTestRule, timeMillis: Int = Delays.RECOMPOSE) {
```

Comment:
Actually I kind of like the sound of `delayAndRecompose`. `wait` sounds like you're waiting for something in particular, `delay` is just "hold on a bit"

Reply:
I agree.
```
delayAndRecompose(composeRule, Delays.CONNECT_SERVER)
```
The lining up of `delayAndRecompose` taking a second argument with the form `Delays.XYZ` looks good.

Reply:
Having seen it in multiple contexts I'm happy with `delayAndRecompose`. It has the same general feeling as `wait` while being more specific and accurate to what's taking place.